keps/173-headless-service-per-replica/README.md

```diff
@@ -0,0 +1,328 @@
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a user running a large scale LWS job, I want to have each leader-worker group with its own service.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### API
+
+```golang
+type LeaderWorkerSetSpec struct {
+ // SubdomainPolicy determines the policy that will be used when creating
+ // the headless service
+ SubdomainPolicy SubdomainPolicy `json:"subdomainPolicy,omitempty"`
+}
+
+type NetworkConfig struct {
+  SubdomainPolicy SubdomainPolicy
+}
+type SubdomainPolicy string
+const (
+  // SubdomainShared will create a single headless service that all replicas 
+  // will share
+  SubdomainShared SubdomainPolicy = "Shared"
+  // SubdomainUniquePerReplica will create a headless service for each
+  // leader-worker group
+  SubdomainUniquePerReplica SubdomainPolicy = "UniquePerReplica"
+)
+```
+
+### Implementation
+
+With HeadlessServicePerReplica true, we will create a headless service per replica.
+The name of the service will be `{lws.Name}-{replicaIndex}`.
+Creation of the services will be in a loop of replicas. The selector of each
+
+With SubdomainPolicy set to SubdomainUniquePerReplica, we will create a headless service
+per replica. The name of the service will be `{lws.Name}-{replicaIndex}`. The selector of
+the headless service will be the group index.
+
+```golang
+for i := 0; i < lws.Spec.Replicas; i++ {
+  headlessService := corev1.Service{
+    ObjectMeta: metav1.ObjectMeta{
+      Name:      fmt.Sprintf("%s-%s", lws.Name, i),
+      Namespace: lws.Namespace,
+    },
+    Spec: corev1.ServiceSpec{
+      ClusterIP: "None", // defines service as headless
+      Selector: map[string]string{
+        leaderworkerset.GroupIndexLabelKey: i,
+      },
+      PublishNotReadyAddresses: true,
+    },
+  }
+}
```

Comment:
how to assign the generated pods to these headless service?

Reply:
It uses the GroupIndexLabelKey as selector, so that way all pods that are part of the same replica are selected to the same headless service

Reply:
we need to add the details how pods will be selected by which headless service as well